### PR TITLE
Enable resuming with `weights.pth` saved by visual prompting

### PR DESCRIPTION
**Pull request by sungchul1**
### Summary

https://github.com/openvinotoolkit/training_extensions/pull/2274#discussion_r1255219663
For this, it was needed to save metas for pytorch lightning in `weights.pth`.
It was implemented by loading best checkpoint saved in `Trainer.checkpoint_callback` and saving metas in `weights.pth`.

### How to test

```python
$ tox -vv -e unittest-all-py310-pt1 -- tests/unit/algorithms/visual_prompting/adapters/pytorch_lightning/config/test_visual_prompting_config.py
=============================================================== 7 passed, 1 warning in 0.72s ===============================================================
Name                                                                                                                Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------------------------------------------------------------------------------
src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/config/visual_prompting_config.py                       43      7    84%   56, 102-103, 106, 111-113


$ tox -vv -e unittest-all-py310-pt1 --  tests/unit/algorithms/visual_prompting/tasks/test_inference.py
============================================================= 31 passed, 17 warnings in 24.33s =============================================================
Name                                                                                                                Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------------------------------------------------------------------------------
src/otx/algorithms/visual_prompting/tasks/inference.py                                                                213     28    87%   96, 115-123, 178, 218-219, 224, 337, 347, 352, 387, 434, 438-447, 454-455, 459-460


$ tox -vv -e unittest-all-py310-pt1 --  tests/unit/algorithms/visual_prompting/tasks/test_train.py
============================================================== 1 passed, 10 warnings in 2.22s ==============================================================
Name                                                                                                                Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------------------------------------------------------------------------------
src/otx/algorithms/visual_prompting/tasks/train.py                                                                     41      4    90%   88-89, 96-97


$ tox -vv -e tests-all-py310-pt1 -- tests/integration/cli/visual_prompting/test_visual_prompting.py
=================================================== 8 passed, 2 skipped, 2 warnings in 108.48s (0:01:48) ===================================================

$ tox -vv -e tests-all-py310-pt1 -- tests/e2e/cli/visual_prompting/test_visual_prompting.py
======================================================== 10 passed, 2 warnings in 274.07s (0:04:34) ========================================================
```

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algorithms/visual_prompting/tasks/inference.py [hunk 1]
```diff
@@ -191,29 +186,31 @@ def get_model(config: DictConfig, state_dict: Optional[OrderedDict] = None):
...
+            # If resuming, pass this part to load checkpoint in Trainer
+            logger.info(f"To resume {otx_model.model_adapters.get('path')}, the checkpoint will be loaded in Trainer.")
+
         else:
-            # pytorch checkpoint saved by otx
+            # Load state_dict
             buffer = io.BytesIO(otx_model.get_data("weights.pth"))
             model_data = torch.load(buffer, map_location=torch.device("cpu"))
-            if model_data.get("model", None) and model_data.get("config", None):
+            if model_data.get("state_dict", None) and model_data.get("pytorch-lightning_version", None):
```
Comment (eunwoosh) [inline src/otx/algorithms/visual_prompting/tasks/inference.py:197]: It's just a question. why is `pytorch-lightning_version` checked?

File: src/otx/algorithms/visual_prompting/tasks/inference.py [hunk 2]
```diff
@@ -191,29 +186,31 @@ def get_model(config: DictConfig, state_dict: Optional[OrderedDict] = None):
...
+                # Load state_dict from pytorch lightning checkpoint or weights.pth saved by visual prompting task
+                state_dict = model_data["state_dict"]
+
+            elif model_data.get("model", None) and model_data.get("config", None):
+                # Load state_dict from checkpoint saved by otx other tasks
```
Comment (eunwoosh) [inline src/otx/algorithms/visual_prompting/tasks/inference.py:205]: Is it possible to load a checkpoint from other tasks in visual prompting?

File: src/otx/algorithms/visual_prompting/tasks/inference.py [hunk 3]
```diff
@@ -406,11 +403,7 @@ def model_info(self) -> Dict:
         Returns:
            Dict: Model info.
         """
-        return {
-            "model": self.model.state_dict(),
-            "config": self.get_config(),
-            "version": self.trainer.logger.version,
-        }
+        return torch.load(self._model_ckpt, map_location="cpu")
```
Comment (eunwoosh) [inline src/otx/algorithms/visual_prompting/tasks/inference.py:412]: How about adding a code to deal with an edge case where `self._model_ckpt` is None?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work! I left some questions. Please take a look.

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:197]: Just to check if `model_data` is from pytorch lightning checkpoint.
There are many metas such as `epoch`, `state_dict`, `loop`, ...
I chose two of them, `state_dict` and `pytorch-lightning_version`.

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:205]: Loading weights is possible, but resuming is not.
I planned that `weights.pth` saved by visual prompting can be used for resuming, not other tasks.
Do you think it is required?

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 1 above)
Comment (eunwoosh) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:197]: Thanks for reply. Then could you add simple comment to explain why it's checked?

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 2 above)
Comment (eunwoosh) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:205]: Ah actually I thought that it's impossible to load other tasks' model weight in visual prompting. Could you explain one example for that?

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:412]: Done.

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:205]: Currently, we cannot test it because there is no task supporting ViT-B as backbone AFAIK.
But, for example, segmentation task saves `weights.pth` with metas, `model`, `config`, `labels`, and `VERSION`.
I think if we have ViT-B weights trained on other tasks, we can use its weights as pretrained weights.
image

Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:205]: If you think it is unnecessary function for now, it would be better to remove this part.

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:197]: It is just to check if `model_data` has `pytorch-lightning_version`, its value is not important.
Other metas can be checked instead of `pytorch-lightning_version`.

Comment (eunwoosh) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:197]: Sorry for confusion. I just want to add a simple comment at that code to give explanation to other developers for better understanding.

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 2 above)
Comment (eunwoosh) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:205]: I think it's better to leave it as it is. Thanks for explanation.

**Review by jaegukhyun — DISMISSED** [review]

File: src/otx/algorithms/visual_prompting/tasks/inference.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/inference.py:197]: Done.

**Review by eunwoosh — APPROVED** [review]

**Review by jaegukhyun — APPROVED** [review]
